bundles/pcmjava/tools.vitruv.applications.pcmjava.common/src/tools/vitruv/applications/pcmjava/common/pcm2java/Pcm2JavaCommon.reactions

```diff
@@ -177,38 +180,138 @@ routine changeSystemImplementationName(pcm::System system) {
+		 val encapsulatedComponentJavaClass = retrieve java::Class corresponding to assemblyContext.encapsulatedComponent__AssemblyContext
+		 require absence of java::Field corresponding to assemblyContext
+	 }
+	 action {
+		call {
+			val foundField = compositeComponentJavaClass.members.filter(Field).findFirst[name == assemblyContext.entityName]
+			if (foundField === null) {
+				createAssemblyContextField(assemblyContext, compositeComponentJavaClass, encapsulatedComponentJavaClass)
+			} else {
+				addAssemblyContextFieldCorrespondence(assemblyContext, foundField)
+			}
+		}
+	 }
+}
+
+routine createAssemblyContextField(pcm::AssemblyContext assemblyContext, java::Class compositeComponentJavaClass, java::Class encapsulatedComponentJavaClass) {
+	action {
+		val assemblyContextField = create java::Field and initialize {
+			createPrivateField(assemblyContextField, createNamespaceClassifierReference(encapsulatedComponentJavaClass), assemblyContext.entityName)
+			compositeComponentJavaClass.members += assemblyContextField
+		}
+		add correspondence between assemblyContextField and assemblyContext
+	}
+}
+
+routine addAssemblyContextFieldCorrespondence(pcm::AssemblyContext assemblyContext, java::Field javaField) {
+	action {
+		add correspondence between javaField and assemblyContext
+	}
 }
 
-// TODO HK This is ugly...
-routine addAssemblyContextToComposedStructure(pcm::ComposedStructure composedStructure, pcm::AssemblyContext assemblyContext) {
+routine createOrFindAssemblyContextConstructor(pcm::AssemblyContext assemblyContext, pcm::ComposedStructure composedStructure) {
 	match {
 		val compositeComponentJavaClass = retrieve java::Class corresponding to composedStructure
 		val encapsulatedComponentJavaClass = retrieve java::Class corresponding to assemblyContext.encapsulatedComponent__AssemblyContext
+		require absence of java::Constructor corresponding to assemblyContext
 	}
 	action {
-		val assemblyContextField = create java::Field and initialize {
-			val TypeReference typeRef = createNamespaceClassifierReference(encapsulatedComponentJavaClass);
-			createPrivateField(assemblyContextField, typeRef, assemblyContext.entityName);
+		call {
+			val foundConstructor = compositeComponentJavaClass.members.filter(Constructor).findFirst[name == compositeComponentJavaClass.name]
+			if (foundConstructor === null) {
+				createAssemblyContextConstructor(assemblyContext, compositeComponentJavaClass)
+			} else {
+				val foundCall = foundConstructor.statements.filter(ExpressionStatement).map[expression].filter(AssignmentExpression).map[value].filter(NewConstructorCall).head
+				if (foundCall === null) {
+					createAssemblyContextStatement(assemblyContext, foundConstructor, compositeComponentJavaClass)
+				} else{
+					addAssemblyContextConstructorCorrespondence(assemblyContext, foundConstructor, foundCall)
+				}
+			}
 		}
-		val newConstructorCall = create java::NewConstructorCall
-		val contextClassImport = create java::ClassifierImport
+	}
+}
```

Comment:
This routine covers a special case, where a Java constructor already exists, but it is empty as it was created from the signature information of a UML constructor. Therefore the constructor needs to be located but the following content needs to be created:
```java
this.testAssemblyContextField = new TestComponent2Impl();
```